Add documentation for TS source (#134373)
elastic/elasticsearch: docs/reference/query-languages/esql/_snippets/commands/layout/ts.md

```yaml {applies_to}
serverless: tech-preview
stack: tech-preview 9.2.0
```

The `TS` command is similar to the `FROM` source command,
with the following key differences:

 - Targets only [time-series indices](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md)
 - Propagates the metadata fields [`_tsid`](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md#tsid)
   and `@timestamp` to the next operators,
   even if they are not mentioned explicitly
 - Enables the use of time-series aggregation functions inside the
   [STATS](/reference/query-languages/esql/commands/stats-by.md) command

> The `TS` command is expected to be used in conjunction with the `STATS` command

Reviewer: ```suggestion
The `TS` command is intended to be used in conjunction with the `STATS` command
```

to perform time-series analysis. `STATS` behaves polymorphically in this context,
enabling the use of time-series aggregation functions such as `last_over_time()`,
or `rate`. These functions are implicitly evaluated per `_tsid`
(per time-series), with their results then aggregated per grouping bucket using
a secondary aggregation function. More concretely, consider the following query:

```esql
TS metrics
  | WHERE @timestamp >= now() - 1 hour
  | STATS total_rate=SUM(RATE(search_requests)) BY hourly=TBUCKET(1 hour), host
```

This query calculates the total rate of search requests (tracked through
counter `search`) per host-hour. Here, the `rate()` function is first
applied per `_tsid` and hourly time bucket, with the results then summed per
host and hourly bucket, as each host value may map to many `_tsid` values
(`_tsid` is calculated on all dimension values).

::::{note}

> Metric fields are part of the time-series definition, so there is no point to
> include null metric values in these queries. Therefore, null metric values get
> filtered out implicitly under the `TS` command.

Reviewer: ```suggestion
Metric fields are part of the time-series definition, so you don't need to
include null metric values in these queries. Null metric values are
filtered out implicitly when using the `TS` command.
```

```yaml {applies_to}
serverless: preview
stack: preview 9.2.0
```

**Brief description**

The `TS` source command is similar to the [`FROM`](/reference/query-languages/esql/commands/from.md)
source command, with the following key differences:

 - Targets only [time-series indices](docs-content://manage-data/data-store/data-streams/time-series-data-stream-tsds.md)

>  - Enables the use of time-series aggregation functions inside the

Reviewer 2: ```suggestion
 - Enables the use of [time-series aggregation functions](/reference/query-languages/esql/functions-operators/time-series-aggregation-functions) inside the
```

`fields`
:   A comma-separated list of [metadata fields](/reference/query-languages/esql/esql-metadata-fields.md) to retrieve.

**Description**

The `TS` source command enables time series semantics and adds support for
time series aggregation functions to the `STATS` command, such as
[`AVG_OVER_TIME()`](/reference/query-languages/esql/functions-operators/time-series-aggregation-functions.md#esql-avg_over_time),
or [`RATE`](/reference/query-languages/esql/functions-operators/time-series-aggregation-functions.md#esql-rate).
These functions are implicitly evaluated per time series, then aggregated by group using a secondary aggregation
function. For example:

```esql
TS metrics
  | WHERE @timestamp >= now() - 1 hour
  | STATS SUM(RATE(search_requests)) BY TBUCKET(1 hour), host
```

This query calculates the total rate of search requests (tracked by the `search_requests` counter) per host and hour. The `RATE()` 
function is applied per time series in hourly buckets. These rates are summed for each

> host and hourly bucket (since each host can map to multiple time series).

Reviewer 3: Sounds like a good plan, thank you Kostas!

apply to downsampled data, with the same semantics as for raw data.

::::{note}
If a query is missing an inner (time series) aggregation function,
[`LAST_OVER_TIME()`](/reference/query-languages/esql/functions-operators/time-series-aggregation-functions.md#esql-last_over_time)
is assumed and used implicitly. For instance, the following two queries are
equivalent, returning the average of the last memory usage values per time series:

```esql
TS metrics | STATS AVG(memory_usage)

TS metrics | STATS AVG(LAST_OVER_TIME(memory_usage))
```

To calculate the average memory usage across per-time-series averages, use
the following query:

```esql
TS metrics | STATS AVG(AVG_OVER_TIME(memory_usage))
```
::::

Use regular (non-time-series) [aggregation functions](/reference/query-languages/esql/functions-operators/aggregation-functions.md),
such as `SUM()`, as outer aggregation functions. Using a time series aggregation
in combination with an inner function causes an error. For example, the
following query is invalid:

```esql
TS metrics | STATS AVG_OVER_TIME(RATE(memory_usage))
```

::::{note}
A time series aggregation function must be wrapped inside a

> regular aggregation function. For instance, the following query is invalid:

Author: Done.

is assumed and used implicitly. For instance, the following two queries are
equivalent, returning the average of the last memory usage values per time-series:

```esql
TS metrics | STATS AVG(memory_usage)

TS metrics | STATS AVG(LAST_OVER_TIME(memory_usage))
```

Calculating the average memory usage across per-time-series averages requires
the following query:

```esql
TS metrics | STATS AVG(AVG_OVER_TIME(memory_usage))
```
::::

Use regular (non-time-series) [aggregation functions](/reference/query-languages/esql/functions-operators/aggregation-functions.md),
such as `SUM()` as outer aggregation functions. Using a time-series aggregation,
in combination with an inner function, leads to an error. For instance, the
following query is invalid:

```esql
TS metrics | STATS AVG_OVER_TIME(RATE(memory_usage))
```

::::{note}
It's currently required to wrap a time-series aggregation function inside a
regular aggregation function. For instance, the following query is invalid:

```esql
TS metrics | STATS RATE(search_requests)
```
::::

**Best practices**

- Avoid mixing aggregation functions on different metrics in the same query, to

>   avoid interference between different time-series. For instance, if one metric

Author: Applied, thanks!

  is missing values for a given time-series, the aggregation function
  may return null for a given combination of dimensions, which may lead to a
  null result for that group if the secondary function returns null on a null
  arg. More so, null metric filtering is more efficient when a query includes
  a single metric.

> - Prefer the `TS` command for aggregations on time-series data. `FROM` is still
>   applicable, e.g. to list document contents, but it's not optimized to process
>   time-series data efficiently. More so, the  `TS` command can't be combined
>   with certain operation such as [`FORK`](/reference/query-languages/esql/commands/fork.md),
>   before the `STATS` command is applied. That said, once `STATS` is applied, its
>   tabular output can be further processed as applicable, in line with regular
>   ES|QL processing.

Reviewer: ```suggestion
- Use the `TS` command for aggregations on time series data, rather than `FROM`. The `FROM` command is still available (for example, for listing document contents), but it's not optimized for time series data.
- The `TS` command can't be combined with certain operations (such as [`FORK`](/reference/query-languages/esql/commands/fork.md)) before the `STATS` command is applied. Once `STATS` is applied, you can process the tabular output with any applicable ES|QL operations.
```

> - Include a time range filter on `@timestamp`, to prevent scanning
>   unnecessarily large data volumes.

Reviewer: ```suggestion
- Add a time range filter on `@timestamp` to limit the data volume scanned and improve query performance.
```
